File: sources/iTermProcessCollection.m
//
//  iTermProcessCollection.m
//  iTerm2
//
//  Created by George Nachman on 4/30/17.
//
//

#import "iTermProcessCollection.h"

#import "NSArray+iTerm.h"
#import "NSObject+iTerm.h"
#import "NSStringITerm.h"

#import "iTerm2SharedARC-Swift.h"

@implementation iTermProcessCollection {
    NSMutableDictionary<NSNumber *, iTermProcessInfo *> *_processes;
    id<iTermProcessDataSource> _dataSource;
}

- (instancetype)initWithDataSource:(id<iTermProcessDataSource>)dataSource {
    self = [super init];
    if (self) {
        _dataSource = dataSource;
        _processes = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (NSString *)treeString {
    return [[_processes.allValues mapWithBlock:^id(iTermProcessInfo *anObject) {
        return [anObject treeStringWithIndent:@""];
    }] componentsJoinedByString:@"\n"];
}

- (iTermProcessInfo *)addProcessWithProcessID:(pid_t)processID
                              parentProcessID:(pid_t)parentProcessID {
    iTermProcessInfo *info = [[iTermProcessInfo alloc] initWithPid:processID
                                                              ppid:parentProcessID
                                                        collection:self
                                                        dataSource:_dataSource];
    _processes[@(processID)] = info;
    return info;
}

- (void)commit {
    [_processes enumerateKeysAndObjectsUsingBlock:^(NSNumber * _Nonnull processID, iTermProcessInfo * _Nonnull info, BOOL * _Nonnull stop) {
        info.parent = self->_processes[@(info.parentProcessID)];
        [info.parent addChildWithProcessID:info.processID];
    }];
}

- (iTermProcessInfo *)infoForProcessID:(pid_t)processID {
    return _processes[@(processID)];
}

@end

File: sources/ProcessInfo/iTermProcessCollection.swift
//
//  iTermProcessCollection.swift
//  iTerm2SharedARC
//
//  Created by George Nachman on 7/12/22.
//

import Foundation

@objc(iTermProcessCollection)
class ProcessCollection: NSObject, ProcessCollectionProvider {
    private var processes = [pid_t: iTermProcessInfo]()
    private let dataSource: ProcessDataSource

    @objc(initWithDataSource:)
    init(dataSource: ProcessDataSource) {
        self.dataSource = dataSource
    }

    @objc var processIDs: [pid_t] {
        return Array(processes.keys)
    }

    var treeString: String {
        return processes.values.map {
            $0.tree(indent: "")
        }.joined(separator: "\n")
    }

    @discardableResult
    func addProcess(withProcessID pid: pid_t, parentProcessID ppid: pid_t) -> iTermProcessInfo {
        let info = iTermProcessInfo(processID: pid,
                                    parentProcessID: ppid,
                                    collection: self,
                                    dataSource: dataSource)
        processes[pid] = info
        return info
    }

    func commit() {
        for info in processes.values {
            info.parent = processes[info.parentProcessID]
            info.parent?.addChild(pid: info.processID)
        }
    }

    func info(forProcessID pid: pid_t) -> iTermProcessInfo? {
        return processes[pid]
    }
}
